src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs

```diff
@@ -57,6 +57,59 @@ public override GeneratorDriver TransformGeneratorDriver(GeneratorDriver generat
             }
         }
 
+        internal sealed class TouchDocumentsAction : TranslationAction
+        {
+            private readonly ImmutableArray<DocumentState> _newStates;
+
+            private TouchDocumentsAction(
+                ProjectState oldProjectState,
+                ProjectState newProjectState,
+                ImmutableArray<DocumentState> newStates)
+                : base(oldProjectState, newProjectState)
+            {
+                _newStates = newStates;
+            }
+
+            public static TranslationAction Create(
+                ProjectState oldProjectState,
+                ProjectState newProjectState,
+                ImmutableArray<DocumentState> newStates)
+            {
+                // Special case when we're only updating a single document.  This case can be optimized more, and
+                // corresponds to the common case of a single file being edited.
+                return newStates.Length == 1
+                    ? new TouchDocumentAction(oldProjectState, newProjectState, oldProjectState.DocumentStates.GetRequiredState(newStates[0].Id), newStates[0])
+                    : new TouchDocumentsAction(oldProjectState, newProjectState, newStates);
```

Comment:
i don't love this.  but there is special logic in TouchDocumentAction (not plural).  It's unclear to me how to maintain that logic here.  So i'm solving it in this fashion.